common/base/include/claraparabricks/genomeworks/utils/device_preallocated_allocator.cuh

```diff
@@ -163,18 +164,14 @@ private:
     /// \param associated_streams on deallocation this memory block is guaranteed to live at least until all previously scheduled work in these streams has finished
     /// \return cudaSuccess if allocation was successful, cudaErrorMemoryAllocation otherwise
     cudaError_t allocate_memory_block(void** ptr,
-                                      size_t bytes_needed,
+                                      const size_t bytes_needed,
                                       const std::vector<cudaStream_t>& associated_streams)
     {
-        *ptr = nullptr;
-
         if (free_blocks_.empty())
         {
```

Comment:
It does not. It is either set to `nullptr` if returning after an error or to the correct value in returning successfully.
I agree that setting it to `nullptr` reduces the chance of forgetting to set it to `nullptr` in case of an error.
@ahehn-nv what do you think?

Reply:
I'm not sure I follow your chain of argument. I think you should put it to `nullptr`, since you return an error.

When reading the change yesterday, I was wondering if you shouldn't leave the `*ptr = nullptr` at the top of the function to have it initialized to a sane value no matter what happens. Then also in case of an exception from std::list, the user would get an initialized ptr from the function. However, I came to the conclusion that it makes sense not to set it at the top, but only when you return an error. Because then the function has no external effects in case of an exception.

Reply:
Yes, that was my idea as well, but @atadkase's question got me thinking that if in the future someone was to change the function and add another error-return they might not realize that `ptr` has to be set to `nullptr`. It's obvious that they have to specify return error code, compiler won't let them get away without it, but it would be easy to overlook `ptr`.
This way we would say that the default value is `nullptr` and one has to change it otherwise. I'm just thinking if this is a bit too much safety

Reply:
You mean, like you did here by returning `cudaErrorMemoryAllocation` without setting `*ptr = nullptr`? ;)

I would make the decision solely on the intended behavior and not based on what bugs future changes could introduce by not respecting this behavior. If somebody modifies this function and the intended behavior or contract of the function is well documented, it is up to them to know what they are doing.
So you might be extra clear in the documentation and change the "ptr on return pointer to allocated memory, nullptr if allocation was not successful" to "*ptr is changed to point to allocated memory if the allocation was successful. In case the function returns an error, *ptr is set to nullptr."

That being said, I think it is equally fine to say "this function will initialize to `*ptr = nullptr` and set `*ptr to the allocated memory iff the allocation was successful".

Reply:
> You mean, like you did here by returning cudaErrorMemoryAllocation without setting *ptr = nullptr? ;)
Exactly 🤣 

Ok. I'll leave it as it is then

Reply:
:smile:

By the way: good catch @atadkase !

Reply:
Was it a conscious decision to not set it to `nullptr` at all now? :laughing:

Reply:
I thought that's what we agreed on in the end. Not setting it to `nullptr` at the beginning because the developer should read the documentation and set `ptr` to the correct value before every return.

Reply:
Oh no, that was a misunderstanding then.
This is also a possibility and I would choose it if the function threw an exception on error.
However, here I would set it to `nullptr` in case of an error - just because the program will fail reliably even if a user does not check the returned error.
In the discussion earlier I was talking about setting it to `nullptr` in any case except returning success (e.g. exception in std::list) vs. setting it only to `nullptr` in case of returning an error.

Reply:
Ok, now I see the missing set-to-nullptr. Done and added tests

Reply:
Thanks Milos! It looks good to me.